packages/material-ui/src/Unstable_TrapFocus/Unstable_TrapFocus.js

```diff
@@ -47,6 +59,24 @@ function Unstable_TrapFocus(props) {
     nodeToRestore.current = getDoc().activeElement;
   }
 
+  const onSentinelFocus = React.useCallback(() => {
+    const focusRadios = rootRef.current.querySelectorAll('input:checked');
+    const isShiftTab = Boolean(lastEvent.current?.shiftKey && lastEvent.current?.keyCode === 9);
+    const selectors = [...focusSelectorsRoot, ...focusSelectors].filter(Boolean);
+    const focusChildren = rootRef.current.querySelectorAll(selectors.join(', '));
```

Comment:
This is overly simplified and I don't see how we could possible determine if an element is focusable by reading the DOM (in a performant way).

What is already missing:
- filter actually disabled elements
- filter expressly inert elements
- apply tabindex order
- filter elements that are not in sequential focus order according to the user-agent (this seems impossible)

This assumes that determining what element is tabbable is something we can do by reading the DOM. However, what elements are tabbable (in sequential focus order) can be determined by the user agent (https://html.spec.whatwg.org/multipage/interaction.html#sequentially-focusable). We can ask if an element is in sequential focus order by e.g. checking tag names or tabIndex. But returning "No" from that question does not imply that the element is not tabbable. Even if we answer yes then the user agent might still reject it (e.g. hidden elements).

Reply:
Agreed.  I'll add the check for disabled and hidden elements.  I'll also review some other implementations to see if there are other attributes we should consider.

After the first, or last tabbable element is focused I am letting the browser control the sequence.  I only coerce focus when the focus lands on a sentinel.